File: vnext/Playground/PlaygroundCPP/SampleModule.cpp
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

#include "pch.h"
#include "SampleModule.h"
#include <winrt\Windows.Data.Json.h>

using namespace winrt;
using namespace Microsoft::ReactNative;
using namespace Windows::Foundation;
using namespace Windows::Foundation::Collections;

namespace winrt::Playground::implementation {
hstring SampleModule::Name() {
  return L"SampleModule";
}

IMapView<hstring, IInspectable> SampleModule::Constants() {
  auto constants = single_threaded_map<hstring, IInspectable>();
  std::wstring value = L"\"b\"";
  constants.Insert(L"a", box_value(L"\"b\""));
  return constants.GetView();
}

IMapView<hstring, MethodDelegate> SampleModule::Methods() {
  auto delegates = single_threaded_map<hstring, MethodDelegate>();
  delegates.Insert(L"method1", [](IVectorView<IInspectable> const &args) {
    // no op
    OutputDebugStringW(L"SampleModule.method1()\n");
  });

  return delegates.GetView();
}

void SampleModule::method3(
    IVectorView<IInspectable> const &args,
    MethodDelegate const &callback) {

  if (args.Size() < 1)
    throw hresult_invalid_argument();

  // At the moment the input args are packed into a single JSON-formatted
  // string.
  auto input = args.GetAt(0);
  auto inputValue = input.as<IPropertyValue>();
  assert(inputValue.Type() == PropertyType::String);
  auto inputString = input.as<IPropertyValue>().GetString();
  OutputDebugStringW(L"SampleModule.method3(");
  OutputDebugStringW(inputString.c_str());
  OutputDebugStringW(L")\n");

  // Invoke the callback with some arguments. Using JSON isn't required.  We
  // could have used other basic types instead (e.g. int, double, float, bool,
  // etc.)
  auto text = L"{ \"result\" : true }";
  auto json = Windows::Data::Json::JsonObject::Parse(text);
  auto callbackArgs = single_threaded_vector<IInspectable>({json}).GetView();

  callback(callbackArgs);
}

IMapView<hstring, MethodWithCallbackDelegate>
SampleModule::MethodsWithCallback() {
  auto delegatesWithCallback =
      single_threaded_map<hstring, MethodWithCallbackDelegate>();
  delegatesWithCallback.Insert(
      L"method2",
      [](IVectorView<IInspectable> const &args,
         MethodDelegate const &callback) {
        OutputDebugStringW(L"SampleModule.method2()\n");

        // Build up a complex object to pass back as arguments when invoking
        // the callback.  They'll automatically be converted to an equivalent
        // type in JavaScript when the JS function is called.
        auto integerArray = PropertyValue::CreateInt32Array(
            std::array<int32_t, 5>{0, 1, 2, 3, 4});

        auto floatArray = PropertyValue::CreateSingleArray(
            std::array<float_t, 3>{0.02f, 1.0f, 0.42f});

        auto stringArray = PropertyValue::CreateStringArray({L"a", L"bb"});

        auto mixedArray = PropertyValue::CreateInspectableArray(
            std::array<IInspectable, 2>{box_value(L"First"), box_value(0x48)});

        auto nestedArray = PropertyValue::CreateInspectableArray(
            std::array<IInspectable, 2>{mixedArray, box_value(L"Last")});

        auto callbackArgs = single_threaded_vector<IInspectable>({box_value(84),
                                                                  integerArray,
                                                                  floatArray,
                                                                  stringArray,
                                                                  nestedArray});

        callback(callbackArgs.GetView());
      });

  delegatesWithCallback.Insert(
      L"method3", MethodWithCallbackDelegate(this, &SampleModule::method3));

  return delegatesWithCallback.GetView();
}

IMapView<hstring, MethodWithPromise> SampleModule::MethodsWithPromise() {
  auto methods = single_threaded_map<hstring, MethodWithPromise>();

  methods.Insert(
      L"method4",
      [](IVectorView<IInspectable> const &args, Promise const &promise) {
        //
        try {
          promise.Resolve(box_value(true));
        } catch (std::exception e) {
          hstring what = to_hstring(e.what());
          promise.Reject(
              L"0x80000000", what, L"stack trace goes here", nullptr);
        }
      });

  return methods.GetView();
}
} // namespace winrt::Playground::implementation

File: vnext/Playground/PlaygroundCS/SampleModule.cs
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

using Microsoft.ReactNative;
using Microsoft.ReactNative.Bridge;
using System;
using System.Linq;
using System.Collections.Generic;
using System.Diagnostics;

namespace Playground
{
  public class SampleModule : INativeModule
  {
    public string Name => nameof(SampleModule);

    public IReadOnlyDictionary<string, object> Constants { get; }
      = new Dictionary<string, object>()
    {
      { "a", "\"b\"" },
    };

    public IReadOnlyDictionary<string, MethodDelegate> Methods { get; }
      = new Dictionary<string, MethodDelegate>()
      {
        { "method1", (args) => { /* fire and forget */ }}
      };

    public IReadOnlyDictionary<string, MethodWithCallbackDelegate> MethodsWithCallback { get; }
      = new Dictionary<string, MethodWithCallbackDelegate>()
      {
        { "method2", (args, callback) => {
          Debug.WriteLine($"SampleModule.method2(${args})");

            // Build up a complex object to pass back as arguments when invoking
            // the callback.  They'll automatically be converted to an equivalent
            // type in JavaScript when the JS function is called.
            var parameters = new List<object> {
              new[] { 0, 1, 2, 3, 4 },
              new[] { 0.02f, 1.0f, 0.42f },
              new[] { "a", "bb" },
              new object[] { new object[] { "First", 0x48 }, "Last" }
            };
            callback(parameters.AsReadOnly());
          }
        },
        {
          // I can refer to the Method3 like this as part of the field initializer because its static.
          // If it wasn't then I'd have to add it later (e.g. as part of the constructor at the earliest).
          "method3", Method3 
        }
      };

    public static void Method3(IReadOnlyList<object> args, MethodDelegate callback)
    {
      // At the moment the input args are packed into a single JSON-formatted string.
      Debug.WriteLine($"SampleModule.method3({args})");

      // Invoke the callback with some arguments. Using JSON isn't required.  We
      // could have used other basic types instead (e.g. int, double, float, bool,
      // etc.)
      string text = @"{ ""result"" : true }";
      // This uses the built-in Windows.Data.Json.JsonObject, but any
      // object whose ToString() returns a json-formatted text could be used
      // (e.g. the Newtonsoft.Json.Linq.JObject). No automatic conversion
      // is done on a string passed directly as arguments. It comes out as a string
      // on the other side and would need to be parsed (e.g. JSON.parse()).
      // Simple... what goes in is what comes out.
      var json = Windows.Data.Json.JsonObject.Parse(text);
      callback(new[] { json });

    }

    public IReadOnlyDictionary<string, MethodWithPromise> MethodsWithPromise { get; }
      = new Dictionary<string, MethodWithPromise>() {
        { "method4", (args, promise) => {
            try
            {
              promise.Resolve(true);
            }
            catch (Exception e)
            {
              promise.Reject("0x80000000", e.Message, e.StackTrace, e);
            }
          }
        }
      };
  }
}
